# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/credentialmanagement/CredentialsContainer.cpp
# Commit: 46af55564501
# Full Hash: 46af555645013a126416a440b658f2969eca4b20
# Author: Benjamin VanderSloot <bvandersloot@mozilla.com>
# Date: 2025-06-06 08:52:03
# Regressor Bug: 1969017
# File Overlap Count: 2
# Description:
#   Bug 1969017 - Refactor FedCM to use a managed actor rather than tagging onto WindowGlobalParent - r=anti-tracking-reviewers,emz
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D251617
# ==============================================================================

diff -r 55c4afa3a46c -r 46af55564501 dom/credentialmanagement/CredentialsContainer.cpp
--- a/dom/credentialmanagement/CredentialsContainer.cpp	Thu Jun 05 14:53:36 2025 +0000
+++ b/dom/credentialmanagement/CredentialsContainer.cpp	Thu Jun 05 14:53:36 2025 +0000
@@ -4,22 +4,17 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-#include "mozilla/Components.h"
-#include "mozilla/CredentialChosenCallback.h"
 #include "mozilla/dom/Credential.h"
 #include "mozilla/dom/CredentialsContainer.h"
 #include "mozilla/dom/FeaturePolicyUtils.h"
-#include "mozilla/dom/IdentityCredential.h"
 #include "mozilla/dom/Promise.h"
-#include "mozilla/dom/Promise-inl.h"
 #include "mozilla/StaticPrefs_dom.h"
 #include "mozilla/StaticPrefs_security.h"
+#include "mozilla/dom/WebIdentityHandler.h"
 #include "mozilla/dom/WebAuthnHandler.h"
 #include "mozilla/dom/WindowGlobalChild.h"
 #include "mozilla/dom/WindowContext.h"
 #include "nsContentUtils.h"
-#include "nsFocusManager.h"
-#include "nsICredentialChooserService.h"
 #include "nsIDocShell.h"
 
 namespace mozilla::dom {
@@ -130,7 +125,7 @@
 }
 
 CredentialsContainer::CredentialsContainer(nsPIDOMWindowInner* aParent)
-    : mParent(aParent), mActiveIdentityRequest(false) {
+    : mParent(aParent) {
   MOZ_ASSERT(aParent);
 }
 
@@ -213,29 +208,18 @@
       return promise.forget();
     }
 
-    if (mActiveIdentityRequest) {
-      promise->MaybeRejectWithInvalidStateError(
-          "Concurrent 'identity' credentials.get requests are not supported."_ns);
+    WebIdentityHandler* identityHandler =
+        mParent->GetOrCreateWebIdentityHandler();
+    if (!identityHandler) {
+      promise->MaybeRejectWithOperationError("");
       return promise.forget();
     }
-    mActiveIdentityRequest = true;
-
-    RefPtr<CredentialsContainer> self = this;
+    if (aOptions.mSignal.WasPassed()) {
+      identityHandler->Follow(&aOptions.mSignal.Value());
+    }
+    identityHandler->GetCredential(aOptions, IsSameOriginWithAncestors(mParent),
+                                   promise);
 
-    promise->AddCallbacksWithCycleCollectedArgs(
-        [](JSContext* aCx, JS::Handle<JS::Value> aValue, ErrorResult& aRv,
-           const RefPtr<CredentialsContainer>& aContainer) {
-          aContainer->mActiveIdentityRequest = false;
-        },
-        [](JSContext* aCx, JS::Handle<JS::Value> aValue, ErrorResult& aRv,
-           const RefPtr<CredentialsContainer>& aContainer) {
-          aContainer->mActiveIdentityRequest = false;
-        },
-        self);
-
-    IdentityCredentialRequestOptions options(aOptions.mIdentity.Value());
-    IdentityCredential::GetCredential(
-        mParent, aOptions, IsSameOriginWithAncestors(mParent), promise);
     return promise.forget();
   }
 
@@ -307,12 +291,14 @@
     return nullptr;
   }
 
-  RefPtr<WindowGlobalChild> wgc = mParent->GetWindowGlobalChild();
-  MOZ_ASSERT(wgc);
+  WebIdentityHandler* identityHandler =
+      mParent->GetOrCreateWebIdentityHandler();
+  if (!identityHandler) {
+    promise->MaybeRejectWithOperationError("");
+    return promise.forget();
+  }
 
-  wgc->SendPreventSilentAccess()->Then(
-      GetCurrentSerialEventTarget(), __func__,
-      [promise] { promise->MaybeResolveWithUndefined(); });
+  identityHandler->PreventSilentAccess(promise);
   return promise.forget();
 }
 